ts/repositories.bzl

```diff
@@ -6,24 +6,85 @@ See https://docs.bazel.build/versions/main/skylark/deploying.html#dependencies
+            You can supply an exact `ts_version` attribute to `rules_ts_dependencies` to bypass this check.""" % ts)
+        version = ts
+
+    if rctx.attr.integrity:
+        integrity = rctx.attr.integrity
+    elif version in TS_VERSIONS.keys():
+        integrity = TS_VERSIONS[version]
+    else:
+        fail("""typescript version {} is not mirrored in rules_ts, is this a real version?
+            If so, you must manually set `ts_integrity`.
+            See documentation on rules_ts_dependencies.""".format(version))
+
+    rctx.download_and_extract(
+        url = [u.format(version) for u in rctx.attr.urls],
+        integrity = integrity,
+    )
+    rctx.symlink(rctx.path(rctx.attr.build_file), "BUILD.bazel")
+
+http_archive_version = repository_rule(
+    implementation = _http_archive_version_impl,
+    attrs = {
+        "integrity": attr.string(doc = "Needed only if the ts version isn't mirrored"),
+        "version": attr.string(doc = "Explicit version for urls placeholder. If provided, the package.json is not read"),
+        "urls": attr.string_list(doc = "URLs to fetch from, each should have a {}-style placeholder"),
+        "build_file": attr.label(doc = "the BUILD file to symlink into the created repository"),
+        "package_json": attr.label(doc = "a path on disk to package.json which may have a version for the package"),
+    },
+)
+
+# WARNING: any additions to this function may be BREAKING CHANGES for users
 # because we'll fetch a dependency which may be different from one that
 # they were previously fetching later in their WORKSPACE setup, and now
 # ours took precedence. Such breakages are challenging for users, so any
 # changes in this function should be marked as BREAKING in the commit message
 # and released only in semver majors.
-def rules_ts_dependencies():
-    "Dependencies needed by users of rules_ts"
+def rules_ts_dependencies(ts_version_from_package_json = None, ts_version = None, ts_integrity = None):
+    """Dependencies needed by users of rules_ts.
+
+    To skip fetching the typescript package, define repository called 'npm_typescript' before calling this.
+
+    Args:
+        ts_version_from_package_json: path to a package.json file which declares an exact typescript version.
+            The version to fetch is read from the file.
+        ts_version: version of the TypeScript compiler.
+            Overrides ts_version_from_package_json.
+            Only needed if ts_version_from_package_json is not set,
+            or the version of typescript in package.json is a semver range.
+        ts_integrity: integrity hash for the npm package.
+            By default, uses values mirrored into rules_ts.
+            You can simply fetch this from the registry:
+            `curl --silent https://registry.npmjs.org/typescript/4.6.3 | jq -r '.dist.integrity'`
+    """
+
+    if not ts_version_from_package_json and not ts_version:
```

Comment:
Should ensure they aren't both set? Would go allow `(not x) != (not y)`?

Reply:
hmm yeah it would probably be a mistake to specify both

Reply:
Actually it looks like specifying both is valid.. so nm?

Reply:
It seems like `ts_version` overrides the package.json, especially if the package.json doesn't have an explicit version (and we don't want to handle `^4.0`?).

Reply:
I don't know what we would do with a semver range. Our goal is to have the same version. We'd need to parse some yarn.lock or package-lock.json or whatever if we want to know what version the editor will actually use.

I made them mutually exclusive.